client/service/src/status_sinks.rs

```diff
@@ -38,56 +46,114 @@ struct YieldAfter<T> {
 impl<T> StatusSinks<T> {
 	/// Builds a new empty collection.
 	pub fn new() -> StatusSinks<T> {
+		let (entries_tx, entries_rx) = tracing_unbounded("status-sinks-entries");
+
 		StatusSinks {
-			entries: FuturesUnordered::new(),
+			inner: Mutex::new(Inner {
+				entries: stream::FuturesUnordered::new(),
+				entries_rx,
+			}),
+			entries_tx,
 		}
 	}
 
 	/// Adds a sender to the collection.
 	///
 	/// The `interval` is the time period between two pushes on the sender.
-	pub fn push(&mut self, interval: Duration, sender: TracingUnboundedSender<T>) {
-		self.entries.push(YieldAfter {
+	pub fn push(&self, interval: Duration, sender: TracingUnboundedSender<T>) {
+		let _ = self.entries_tx.unbounded_send(YieldAfter {
 			delay: Delay::new(interval),
 			interval,
 			sender: Some(sender),
-		})
+		});
 	}
 
-	/// Processes all the senders. If any sender is ready, calls the `status_grab` function and
-	/// pushes what it returns to the sender.
+	/// Waits until one of the sinks is ready, then returns an object that can be used to send
+	/// an element on said sink.
 	///
-	/// This function doesn't return anything, but it should be treated as if it implicitly
-	/// returns `Poll::Pending`. In particular, it should be called again when the task
-	/// is waken up.
-	///
-	/// # Panic
-	///
-	/// Panics if not called within the context of a task.
-	pub fn poll(&mut self, cx: &mut Context, mut status_grab: impl FnMut() -> T) {
+	/// If the object isn't used to send an element, the slot is skipped.
+	pub async fn next(&self) -> ReadySinkEvent<'_, T> {
+		// This is only ever locked by `next`, which means that one `next` at a time can run.
+		let mut inner = self.inner.lock().await;
+		let inner = &mut *inner;
+
 		loop {
-			match Pin::new(&mut self.entries).poll_next(cx) {
-				Poll::Ready(Some((sender, interval))) => {
-					let status = status_grab();
-					if sender.unbounded_send(status).is_ok() {
-						self.entries.push(YieldAfter {
-							// Note that since there's a small delay between the moment a task is
-							// waken up and the moment it is polled, the period is actually not
-							// `interval` but `interval + <delay>`. We ignore this problem in
-							// practice.
-							delay: Delay::new(interval),
-							interval,
-							sender: Some(sender),
-						});
+			// Future that produces the next ready entry in `entries`, or doesn't produce anything if
+			// the list is empty.
+			let next_ready_entry = {
+				let entries = &mut inner.entries;
+				async move {
+					if let Some(v) = entries.next().await {
+						v
+					} else {
+						loop {
```

Comment:
Hopefully not a dumb question: What is this `loop` for?

Reply:
I'm not actually sure whether this loop is necessary. I think `pending!` is like `std::thread::yield_now` and can return.